yarpl/include/yarpl/flowable/PublishProcessor.h

```diff
@@ -0,0 +1,234 @@
+    publishers_.swap(publishers);
+
+    for (const auto& publisher : *publishers) {
+      publisher->terminate();
+    }
+  }
+
+  void subscribe(std::shared_ptr<Subscriber<T>> subscriber) override {
+    auto publisher = std::make_shared<PublisherSubscription>(subscriber, this);
+    // we have to call onSubscribe before adding it to the list of publishers
+    // because they might start emitting right away
+    subscriber->onSubscribe(publisher);
+
+    if (publisher->isCancelled()) {
+      return;
+    }
+
+    auto publishers = tryAddPublisher(publisher);
+
+    if (publishers == kCompletedPublishers) {
+      publisher->onComplete();
+    } else if (publishers == kErroredPublishers) {
+      publisher->onError(std::runtime_error("ErroredPublisher"));
+    }
+  }
+
+  void onSubscribe(std::shared_ptr<Subscription> subscription) override {
+    auto publishers = publishers_.copy();
+    if (publishers == kCompletedPublishers ||
+        publishers == kErroredPublishers) {
+      subscription->cancel();
+      return;
+    }
+
+    subscription->request(credits::kNoFlowControl);
+  }
+
+  void onNext(T value) override {
+    auto publishers = publishers_.copy();
+    DCHECK(publishers != kCompletedPublishers);
+    DCHECK(publishers != kErroredPublishers);
+
+    for (const auto& publisher : *publishers) {
+      publisher->onNext(value);
+    }
+  }
+
+  void onError(folly::exception_wrapper ex) override {
+    auto publishers = kErroredPublishers;
+    publishers_.swap(publishers);
+    DCHECK(publishers != kCompletedPublishers);
+    DCHECK(publishers != kErroredPublishers);
+
+    for (const auto& publisher : *publishers) {
+      publisher->onError(ex);
+    }
+  }
+
+  void onComplete() override {
+    auto publishers = kCompletedPublishers;
+    publishers_.swap(publishers);
+    DCHECK(publishers != kCompletedPublishers);
+    DCHECK(publishers != kErroredPublishers);
+
+    for (const auto& publisher : *publishers) {
+      publisher->onComplete();
+    }
+  }
+
+ private:
+  std::shared_ptr<const PublishersVector> tryAddPublisher(
+      std::shared_ptr<PublisherSubscription> subscriber) {
+    while (true) {
+      auto oldPublishers = publishers_.copy();
+      if (oldPublishers == kCompletedPublishers ||
+          oldPublishers == kErroredPublishers) {
+        return oldPublishers;
+      }
+
+      auto newPublishers = std::make_shared<PublishersVector>();
+      newPublishers->reserve(oldPublishers->size() + 1);
+      newPublishers->insert(
+          newPublishers->begin(),
+          oldPublishers->cbegin(),
+          oldPublishers->cend());
+      newPublishers->push_back(subscriber);
+
+      auto locked = publishers_.lock();
+      if (*locked == oldPublishers) {
+        *locked = newPublishers;
+        return newPublishers;
+      }
+      // else the vector changed so we will have to do it again
+    }
+  }
+
+  void removePublisher(PublisherSubscription* subscriber) {
+    while (true) {
```

Comment:
this can be a lot simpler and likely just as performant: Exclusive lock on `publishers_` and [`std::remove_if`](http://en.cppreference.com/w/cpp/algorithm/remove)